torch/csrc/api/src/optim/adagrad.cpp

```diff
@@ -55,8 +55,12 @@ void AdagradParamState::serialize(torch::serialize::InputArchive& archive) {
 
 /// Adapted from
 /// https://github.com/pytorch/pytorch/blob/master/torch/optim/adagrad.py
-void Adagrad::step() {
+Tensor Adagrad::step(LossClosure closure) {
   NoGradGuard no_grad;
+  Tensor loss = {};
+  if (closure != nullptr) {
+    loss = closure();
```

Comment:
We likely need the equivalent of `with torch.enable_grad():` above this line. I believe `torch::AutoGradMode enable_grad(true);` (example: https://github.com/pytorch/pytorch/blob/82cdd3abae173985c21889522cbe9f7d6dc43dc5/test/cpp/api/autograd.cpp#L483) should work here.